### DISCUSSION TITLE
String interpolation

### DISCUSSION THREAD

**Discussion by borkdude** [clavascript/clavascript#40] [Ideas]
@corasaurus-hex String interpolation is a different issue: this issue is just about generating more efficient code for `(str foo bar)` => `cherry/str(foo,bar)` vs `foo + bar`.
I think string interpolation should also be a built-in feature that compiles directly to JS string interpolation. We can support this using a reader template `#i "foo ${name}"`. I'll make a different issue for that.

_Originally posted by @borkdude in https://github.com/clavascript/clavascript/issues/10#issuecomment-1217531806_

**Comment by lilactown** [comment]
Nitpicking: `#i` is a bit ambiguous, what about `#str`?

**Comment by borkdude** [comment]
Yes, `#i` isn't a good choice. Not sure about `#str`, maybe too general?. What about `#<<`? `<<` was the string interpolation macro in clojure.contrib. Too cryptic? 

```
#<< "hello ${name}"
```

**Comment by borkdude** [comment]
Or `#$ "hello ${name}`: too Perly...

**Comment by lilactown** [comment]
Outside of the reader tag we use, how ought we to handle custom tag functions? Something that would expand to

```javascript
html`<div style="color: red;">hello ${user}!</div>`
```

**Comment by lilactown** [comment]
perhaps we could allow passing in a list, like `#tmpl (html "<div style=\"color: red;\">hello ${user}!</div>")`

**Comment by corasaurus-hex** [comment]
maybe #` ? Some way to indicate that these are backticks?

**Comment by corasaurus-hex** [comment]
oh dang that's not bad @lilactown, I like it

**Comment by corasaurus-hex** [comment]
how does that generalize to bare template strings, though?

**Comment by lilactown** [comment]
i'm a bit ignorant of how the reader works is but i think we could have it support both bare strings as well as lists with a tag function

```clojure
#tmpl "Hello, ${user}!"

#tmpl (html "<div style=\"color: red;\">hello ${user}!</div>")
```

I'm using `tmpl` as a stand in for whatever we decide to use.

**Comment by lilactown** [comment]
Thinking out loud, I'm worried about how we'll handle complex cases than just interpolating a variable. For instance, how should we handle expressions?

```clojure
#tmpl "Hello, ${(:name user)}!"
```

Perhaps we should lean into the tag accepting a collection, and instead of writing the `${}` syntax directly we write lists

```clojure
#tmpl ("Hello, " user-name "!")
#tmpl ("Hello, " (:name user) "!")
#tmpl (html "<div style=\"color: red;\">Hello, " user-name "!</div>")
```

**Comment by borkdude** [comment]
> #tmpl ("Hello, " user-name "!")

Hmm, this isn't any better to me than just using `str`. The point of `"${foo}"` is that it's easier to write.

We could support `#tmpl "Hello ${name}"` and `#tmpl [:html "<pre>${name}</pre>]`.

**Comment by lilactown** [comment]
@borkdude how should we handle arbitrary expressions inside of the template string, like I was talking about above?

**Reply by borkdude** [reply]
```
#tmpl "Hello, ${(:name user)}!"
```

would be compiled to:

``` javascript
"Hello" + (:name user) + "!"
```

**Reply by borkdude** [reply]
Yes, good point, this show why we can't directly compile to JS interpolation but have to parse those strings in the compiler, split them and them compile the sub-expressions.

**Reply by lilactown** [reply]
Right. That's why I suggested the list version with the strings broken out above, to avoid having to parse the strings. The benefit to the user would be that it would compile to efficient interpolation and that you could use template tags easily.

**Reply by borkdude** [reply]
But then it would not be interpolation anymore: it would just be the user splitting things manually and you might as well use `str` then. The interpolation is there for convenience. Typing `(str "(" hello ")")` is very awkward in a Clojure editor, compared to `#tmpl "( ${hello} )"`.

**Reply by lilactown** [reply]
Editors and other tooling will need to add support for the interpolation syntax in order to do syntax highlighting and other static analysis. I'm worried that, while the syntax is nice while writing it, it's not worth the challenges it presents to the reader.

Maybe it could use the interpolation syntax, but bindings have to be provided some way?

```clojure
#tmpl ("Hello, ${name}!" {:name (:name user)})

#tmpl (html "<pre>${name}</pre>" {:name (:name user)})
```

**Reply by lilactown** [reply]
To be up front, I see the differences between this and `str` to be very small - I prefer using `str` over `format` and similar most of the time - and am mostly motivated by the ability to easily use tagged literals like lit-html's `html` tag

**Reply by borkdude** [reply]
@lilactown I'm very sensitive to the argument that static analysis should not be made harder :-) I like your idea of providing a map with arguments, this is quite similar to how Selmer works.

**Reply by borkdude** [reply]
``` clojure
(ns subst
  (:require [clojure.string :as str]))

(defn subst [s m]
  (reduce-kv
   (fn [acc k v]
     (str/replace acc (str "$" (name k)) (str v)))
   s m))

(prn (subst "$num + $num = $sum"
            {:num 1
             :sum (+ 1 1)}))
;; "1 + 1 = 2"
```

I think it doesn't even have to be a reader macro, could just be a function or maybe a macro to optimize constants and then defer to a runtime construct.

**Reply by lilactown** [reply]
I think we'd want it at least to be a special form so that it can output backtics

**Reply by borkdude** [reply]
Oh wait, yeah, we were constructing a compiled form from this, I forgot.
